pkg/util/kubernetes/apiserver/controller_util.go

```diff
@@ -0,0 +1,208 @@
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/DataDog/datadog-agent/pkg/clusteragent/custommetrics"
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/util/kubernetes/apiserver/common"
+	"github.com/DataDog/datadog-agent/pkg/util/kubernetes/autoscalers"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/watermarkpodautoscaler/pkg/apis/datadoghq/v1alpha1"
+)
+
+// NewAutoscalersController returns a new AutoscalersController
+func NewAutoscalersController(client kubernetes.Interface, le LeaderElectorInterface, dogCl autoscalers.DatadogClient) (*AutoscalersController, error) {
+	var err error
+
+	h := &AutoscalersController{
+		clientSet: client,
+		le:        le, // only trigger GC and updateExternalMetrics by the Leader.
+		queue:     workqueue.NewNamedRateLimitingQueue(workqueue.DefaultItemBasedRateLimiter(), "autoscalers"),
+	}
+
+	h.toStore.data = make(map[string]custommetrics.ExternalMetricValue)
+
+	gcPeriodSeconds := config.Datadog.GetInt("hpa_watcher_gc_period")
+	refreshPeriod := config.Datadog.GetInt("external_metrics_provider.refresh_period")
+
+	if gcPeriodSeconds <= 0 || refreshPeriod <= 0 {
+		return nil, fmt.Errorf("tickers must be strictly positive in the AutoscalersController"+
+			" [GC: %d s, Refresh: %d s]", gcPeriodSeconds, refreshPeriod)
+	}
+
+	h.poller = PollerConfig{
+		gcPeriodSeconds: gcPeriodSeconds,
+		refreshPeriod:   refreshPeriod,
+	}
+
+	// Setup the client to process the Ref and metrics
+	h.hpaProc, err = autoscalers.NewProcessor(dogCl)
+	if err != nil {
+		log.Errorf("Could not instantiate the Ref Processor: %v", err.Error())
+		return nil, err
+	}
+
+	datadogHPAConfigMap := custommetrics.GetConfigmapName()
+	h.store, err = custommetrics.NewConfigMapStore(client, common.GetResourcesNamespace(), datadogHPAConfigMap)
+	if err != nil {
+		log.Errorf("Could not instantiate the local store for the External Metrics %v", err)
+		return nil, err
+	}
+	return h, nil
+}
+
+func (h *AutoscalersController) enqueue(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		log.Debugf("Couldn't get key for object %v: %v", obj, err)
+		return
+	}
+	h.queue.AddRateLimited(key)
+}
+
+// RunControllerLoop is the public method to trigger the lifecycle loop of the External Metrics store
+func (h *AutoscalersController) RunControllerLoop(stopCh <-chan struct{}) {
+	h.processingLoop()
+}
+
+// gc checks if any hpas or wpas have been deleted (possibly while the Datadog Cluster Agent was
+// not running) to clean the store.
+func (h *AutoscalersController) gc() {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	log.Infof("Starting garbage collection process on the Autoscalers")
+
+	list, err := h.autoscalersLister.HorizontalPodAutoscalers(metav1.NamespaceAll).List(labels.Everything())
+	if err != nil {
+		log.Errorf("Could not list hpas: %v", err)
```

Comment:
do you want to continue here?

Reply:
no because we also want to GC the WPAs. which is done right after if enabled